RestartSource/Sink not working with AmqpCachedConnectionProvider

After changing to use the `AmqpCachedConnectionProvider` (in order to minimize number of open connections) the `RestartSource` stops working. The Source will be finished with an error when RabbitMQ shuts down.
The code below works, it will recover when RabbitMQ has been restarted. Removing the comments and enabling the `AmqpCachedConnectionProvider` it will just stop the Flow with the error message:

    [ERROR] [03/17/2018 14:46:14.561] [MySpec-akka.actor.default-dispatcher-3] [akka://MySpec/user/StreamSupervisor-0/flow-0-0-ignoreSink] Error during postStop in [akka.stream.alpakka.amqp.AmqpSourceStage@4333f384]: connection is already closed due to connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
    com.rabbitmq.client.AlreadyClosedException: connection is already closed due to connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
	at com.rabbitmq.client.impl.AMQConnection.startShutdown(AMQConnection.java:875)
	at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:991)
	at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:920)
	at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:904)
	at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:896)
	at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.close(AutorecoveringConnection.java:216)
	at akka.stream.alpakka.amqp.AmqpConnectionProvider.release(AmqpConnectionProvider.scala:21)
	at akka.stream.alpakka.amqp.AmqpConnectionProvider.release$(AmqpConnectionProvider.scala:21)
	at akka.stream.alpakka.amqp.AmqpDetailsConnectionProvider.release(AmqpConnectionProvider.scala:52)
	at akka.stream.alpakka.amqp.AmqpCachedConnectionProvider.release(AmqpConnectionProvider.scala:241)
	at akka.stream.alpakka.amqp.AmqpConnectorLogic.postStop(AmqpConnector.scala:80)
	at akka.stream.alpakka.amqp.AmqpConnectorLogic.postStop$(AmqpConnector.scala:77)
	at akka.stream.alpakka.amqp.AmqpSourceStage$$anon$1.postStop(AmqpSourceStage.scala:49)
	at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:564)
	at akka.stream.impl.fusing.GraphInterpreter.afterStageHasRun(GraphInterpreter.scala:545)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:460)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:465)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:560)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:742)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:757)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
	at akka.actor.ActorCell.invoke(ActorCell.scala:496)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

    [ERROR] [03/17/2018 14:46:14.563] [MySpec-akka.actor.default-dispatcher-3] [RestartWithBackoffSource(akka://MySpec)] Restarting graph due to failure
    com.rabbitmq.client.ShutdownSignalException: connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
	at com.rabbitmq.client.impl.AMQConnection.startShutdown(AMQConnection.java:870)
	at com.rabbitmq.client.impl.AMQConnection.shutdown(AMQConnection.java:860)
	at com.rabbitmq.client.impl.AMQConnection.handleConnectionClose(AMQConnection.java:798)
	at com.rabbitmq.client.impl.AMQConnection.processControlCommand(AMQConnection.java:753)
	at com.rabbitmq.client.impl.AMQConnection$1.processAsync(AMQConnection.java:237)
	at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:178)
	at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:111)
	at com.rabbitmq.client.impl.AMQConnection.readFrame(AMQConnection.java:623)
	at com.rabbitmq.client.impl.AMQConnection.access$300(AMQConnection.java:47)
	at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:581)
	at java.base/java.lang.Thread.run(Thread.java:844)


Code:

    import scala.concurrent.duration._
    implicit val mat = ActorMaterializer()

    val (min, max, random) = (1.seconds, 10.seconds, 0.2)

    val connectionProvider =
    //  AmqpCachedConnectionProvider(
      AmqpDetailsConnectionProvider(
        "localhost", 5672
      )
        .withVirtualHost("/")
        .withCredentials(AmqpCredentials("user", "password"))
        .withAutomaticRecoveryEnabled(true)
        .withTopologyRecoveryEnabled(true)
        .withNetworkRecoveryInterval(10)
        .withRequestedHeartbeat(10)
    //    )
    val exchangeDeclaration = ExchangeDeclaration(
      "exchange",
      "topic",
      durable = true
    )


    val source = AmqpSource.committableSource(
      TemporaryQueueSourceSettings(connectionProvider, exchangeDeclaration.name, routingKey = Some("test"))
        .withDeclarations(exchangeDeclaration),
      bufferSize = 1
    )

    RestartSource
      .withBackoff(min, max, random) { () => source }
      .runWith(Sink.ignore)

   
Is this to be expected?

--- a/amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala
+++ b/amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala
@@ -18,7 +18,7 @@ import scala.annotation.tailrec
 @DoNotInherit
 sealed trait AmqpConnectionProvider {
   def get: Connection
-  def release(connection: Connection): Unit = connection.close()
+  def release(connection: Connection): Unit = if (connection.isOpen) connection.close()
 }
 
 /**

--- a/amqp/src/test/java/akka/stream/alpakka/amqp/javadsl/AmqpConnectionProvidersTest.java
+++ b/amqp/src/test/java/akka/stream/alpakka/amqp/javadsl/AmqpConnectionProvidersTest.java
@@ -24,6 +24,14 @@ public class AmqpConnectionProvidersTest {
   }
 
   @Test
+  public void LocalAmqpConnectionReleaseClosedConnectionDoNotError() throws Exception {
+    AmqpConnectionProvider connectionProvider = AmqpLocalConnectionProvider.getInstance();
+    Connection connection1 = connectionProvider.get();
+    connectionProvider.release(connection1);
+    connectionProvider.release(connection1);
+  }
+
+  @Test
   public void AmqpConnectionUriCreatesNewConnection() throws Exception {
     AmqpConnectionProvider connectionProvider = AmqpUriConnectionProvider.create("amqp://localhost:5672");
     Connection connection1 = connectionProvider.get();
@@ -31,6 +39,14 @@ public class AmqpConnectionProvidersTest {
     assertNotEquals(connection1, connection2);
     connectionProvider.release(connection1);
     connectionProvider.release(connection2);
+  }
+
+  @Test
+  public void AmqpConnectionUriReleaseClosedConnectionDoNotError() throws Exception {
+    AmqpConnectionProvider connectionProvider = AmqpUriConnectionProvider.create("amqp://localhost:5672");
+    Connection connection1 = connectionProvider.get();
+    connectionProvider.release(connection1);
+    connectionProvider.release(connection1);
   }
 
   @Test
@@ -44,6 +60,14 @@ public class AmqpConnectionProvidersTest {
   }
 
   @Test
+  public void AmqpConnectionDetailsReleaseClosedConnectionDoNotError() throws Exception {
+    AmqpConnectionProvider connectionProvider = AmqpDetailsConnectionProvider.create("localhost", 5672);
+    Connection connection1 = connectionProvider.get();
+    connectionProvider.release(connection1);
+    connectionProvider.release(connection1);
+  }
+
+  @Test
   public void AmqpConnectionFactoryCreatesNewConnection() throws Exception {
     ConnectionFactory connectionFactory = new ConnectionFactory();
     @SuppressWarnings("unchecked")
@@ -54,6 +78,17 @@ public class AmqpConnectionProvidersTest {
     assertNotEquals(connection1, connection2);
     connectionProvider.release(connection1);
     connectionProvider.release(connection2);
+  }
+
+  @Test
+  public void AmqpConnectionFactoryReleaseClosedConnectionDoNotError() throws Exception {
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    @SuppressWarnings("unchecked")
+    AmqpConnectionProvider connectionProvider = AmqpConnectionFactoryConnectionProvider.create(connectionFactory)
+            .withHostsAndPorts(Pair.create("localhost", 5672));
+    Connection connection1 = connectionProvider.get();
+    connectionProvider.release(connection1);
+    connectionProvider.release(connection1);
   }
 
   @Test

--- a/amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpConnectionProvidersSpec.scala
+++ b/amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpConnectionProvidersSpec.scala
@@ -18,6 +18,13 @@ class AmqpConnectionProvidersSpec extends AmqpSpec {
       connectionProvider.release(connection2)
     }
 
+    "not error if releasing an already closed LocalAmqpConnection" in {
+      val connectionProvider = AmqpLocalConnectionProvider
+      val connection1 = connectionProvider.get
+      connectionProvider.release(connection1)
+      connectionProvider.release(connection1)
+    }
+
     "create a new connection per invocation of AmqpConnectionUri" in {
       val connectionProvider = AmqpUriConnectionProvider("amqp://localhost:5672")
       val connection1 = connectionProvider.get
@@ -25,6 +32,13 @@ class AmqpConnectionProvidersSpec extends AmqpSpec {
       connection1 should not equal connection2
       connectionProvider.release(connection1)
       connectionProvider.release(connection2)
+    }
+
+    "not error if releasing an already closed AmqpConnectionUri" in {
+      val connectionProvider = AmqpUriConnectionProvider("amqp://localhost:5672")
+      val connection1 = connectionProvider.get
+      connectionProvider.release(connection1)
+      connectionProvider.release(connection1)
     }
 
     "create a new connection per invocation of AmqpConnectionDetails" in {
@@ -36,6 +50,13 @@ class AmqpConnectionProvidersSpec extends AmqpSpec {
       connectionProvider.release(connection2)
     }
 
+    "not error if releasing an already closed AmqpConnectionDetails" in {
+      val connectionProvider = AmqpDetailsConnectionProvider(List(("localhost", 5672)))
+      val connection1 = connectionProvider.get
+      connectionProvider.release(connection1)
+      connectionProvider.release(connection1)
+    }
+
     "create a new connection per invocation of AmqpConnectionFactory" in {
       val connectionFactory = new ConnectionFactory()
       val connectionProvider = AmqpConnectionFactoryConnectionProvider(connectionFactory, List(("localhost", 5672)))
@@ -44,6 +65,14 @@ class AmqpConnectionProvidersSpec extends AmqpSpec {
       connection1 should not equal connection2
       connectionProvider.release(connection1)
       connectionProvider.release(connection2)
+    }
+
+    "not error if releasing an already closed AmqpConnectionFactory" in {
+      val connectionFactory = new ConnectionFactory()
+      val connectionProvider = AmqpConnectionFactoryConnectionProvider(connectionFactory, List(("localhost", 5672)))
+      val connection1 = connectionProvider.get
+      connectionProvider.release(connection1)
+      connectionProvider.release(connection1)
     }
   }
 

